lib/main.js

```diff
@@ -6,7 +6,80 @@ import { dirname, extname } from 'path';
 
 let helpers = null;
 let clangFlags = null;
-const regex = /(.+):(\d+):(\d+):(?:{(\d+):(\d+)-(\d+):(\d+)}.*:)? ([\w \\-]+): (.*)/g;
+
+const regex = new RegExp([
+  '^(<stdin>|.+):', // Path, usually <stdin>
+  '(\\d+):(\\d+):', // Base line and column
+  '(?:({.+}):)?', // Range position(s), if present
+  ' ([\\w \\\\-]+):', // Message type
+  ' ([^[\\n\\r]+)', // The message
+  '(?: \\[(.+)\\])?\\r?$', // -W flag, if any
+  '(?:\\r?\\n^ .+$)+', // The visual caret diagnostics, necessary to include in output for fix-its
+  '(?:\\r?\\n^fix-it:".+":', // Start of fix-it block
+  '{(\\d+):(\\d+)-(\\d+):(\\d+)}:', // fix-it range
+  '"(.+)"', // fix-it replacement text
+  '$)?', // End of fix-it block
+].join(''), 'gm');
+
+/**
+ * Given a set of ranges in clangs format, determine the range encompasing all points
+ * @param  {String} ranges The raw range string to parse
+ * @return {Range}        An Atom Range object encompasing all given ranges
+ */
+const parseClangRanges = (ranges) => {
+  const rangeRE = /{(\d+):(\d+)-(\d+):(\d+)}/g;
+  let lineStart;
+  let colStart;
+  let lineEnd;
+  let colEnd;
+
+  let match = rangeRE.exec(ranges);
+  while (match !== null) {
+    const rangeLineStart = Number.parseInt(match[1], 10) - 1;
+    const rangeColStart = Number.parseInt(match[2], 10) - 1;
+    const rangeLineEnd = Number.parseInt(match[3], 10) - 1;
+    const rangeColEnd = Number.parseInt(match[4], 10) - 1;
+    if (lineStart === undefined) {
+      // First match
+      lineStart = rangeLineStart;
+      colStart = rangeColStart;
+      lineEnd = rangeLineEnd;
+      colEnd = rangeColEnd;
+    } else {
+      if (rangeLineStart > lineEnd) {
+        // Higher starting line
+        lineEnd = rangeLineStart;
+        colEnd = rangeColStart;
+      }
+      if (rangeLineEnd > lineEnd) {
+        // Higher ending line
+        lineEnd = rangeLineEnd;
+        colEnd = rangeColEnd;
+      }
+      if (rangeColEnd > colEnd) {
+        // Higher ending column
+        colEnd = rangeColEnd;
+      }
+    }
+    match = rangeRE.exec(ranges);
+  }
+  return [[lineStart, colStart], [lineEnd, colEnd]];
+};
+
+/**
+ * Determine if a given path is open in an existing TextEditor
+ * @param  {String} filePath The file path to search for an editor of
+ * @return {TextEditor | false}      The TextEditor or false if none found
+ */
+const findTextEditor = (filePath) => {
+  const allEditors = atom.workspace.getTextEditors();
+  const matchingEditor = allEditors.find(
+    textEditor => textEditor.getPath() === filePath);
+  if (matchingEditor !== undefined) {
```

Comment:
No need to be super-explicit here as well, could just be `return matchingEditor || false` 😛